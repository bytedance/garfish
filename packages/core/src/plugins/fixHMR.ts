import { interfaces } from '../interface';

// When the main application is updated, the currently active child applications need to rerender.
export function GarfishHMRPlugin() {
  let hasInit = false;
  let isHotUpdate = false;
  return function (Garfish: interfaces.Garfish): interfaces.Plugin {
    return {
      name: 'fix-hmr',
      version: __VERSION__,

      bootstrap() {
        if (hasInit) return;
        hasInit = true;

        let webpackHotUpdateName = 'webpackHotUpdate';
        let webpackHotUpdate = (window as any)[webpackHotUpdateName];

        for (const i in window) {
          if (i.includes('webpackHotUpdate')) {
            webpackHotUpdateName = i;
            webpackHotUpdate = window[i];
          }
        }

        for (const i in window) {
          if (i.includes('webpackHotUpdate')) {
            webpackHotUpdate = window[i];
            webpackHotUpdateName = i;
          }
        }

        if (typeof webpackHotUpdate === 'function') {
          (window as any)[webpackHotUpdateName] = function () {
            isHotUpdate = true;
            return webpackHotUpdate.apply(this, arguments);
          };

          const observer = new MutationObserver(() => {
            if (!isHotUpdate) return;
            isHotUpdate = false;

            Garfish.activeApps.forEach((app) => {
              if (app.mounted) {
                setTimeout(() => {
                  app.display && app.hide();
                  app.show();
                });
              }
            });
          });

          observer.observe(document.documentElement, {
            subtree: true,
            childList: true,
            attributes: true,
          });
        }
      },
    };
  };
}
