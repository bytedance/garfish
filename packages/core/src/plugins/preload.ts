import {
  warn,
  Queue,
  isAbsolute,
  transformUrl,
  callTestCallback,
  safeWrapper,
} from '@garfish/utils';
import { Loader, Manager, TemplateManager } from '@garfish/loader';
import { interfaces } from '../interface';

export const storageKey = '__garfishPreloadApp__';

const isMobile =
  /Android|webOS|iPhone|iPad|iPod|BlackBerry|IEMobile|Opera Mini/i.test(
    navigator.userAgent,
  );

// Using queues, to avoid interference with the normal request
const requestQueue = new Queue();

const isSlowNetwork = () =>
  (navigator as any).connection
    ? (navigator as any).connection.saveData ||
      /(2|3)g/.test((navigator as any).connection.effectiveType)
    : false;

const idleCallback =
  (window as any).requestIdleCallback || window.requestAnimationFrame;

export const requestIdleCallback =
  __TEST__ || typeof idleCallback !== 'function'
    ? window.setTimeout
    : idleCallback;

// Test size, catch mistakes, avoid preload first screen white during parsing error
function safeLoad(
  loader: Loader,
  appName: string,
  url: string,
  isModule: boolean,
  immediately: boolean,
  callback?: (m: Manager) => any,
) {
  requestQueue.add((next) => {
    const throwWarn = (e) => {
      next();
      if (__DEV__) {
        warn(e);
        warn(`Preload failed. "${url}"`);
      }
    };

    const success = ({ resourceManager }) => {
      callback && callback(resourceManager);
      setTimeout(next, 500);
    };

    const loadResource = () => {
      try {
        if (isModule) {
          loader.loadModule(url).then(success, throwWarn);
        } else {
          loader.load({ scope: appName, url }).then(success, throwWarn);
        }
      } catch (e) {
        throwWarn(e);
      }
    };

    // edge
    immediately ? loadResource() : requestIdleCallback(loadResource);
  });
}

export function loadAppResource(
  loader: Loader,
  info: interfaces.AppInfo,
  immediately = false,
) {
  __TEST__ && callTestCallback(loadAppResource, info);
  const fetchUrl = transformUrl(location.href, info.entry);

  safeLoad(loader, info.name, fetchUrl, false, immediately, (manager) => {
    const loadStaticResource = () => {
      if (manager instanceof TemplateManager) {
        const baseUrl = manager.url;
        const jsNodes = manager.findAllJsNodes();
        const linkNodes = manager.findAllLinkNodes();
        const metaNodes = manager.findAllMetaNodes();

        if (jsNodes) {
          jsNodes.forEach((node) => {
            const src = manager.findAttributeValue(node, 'src');
            src &&
              safeLoad(
                loader,
                info.name,
                baseUrl ? transformUrl(baseUrl, src) : src,
                false,
                immediately,
              );
          });
        }
        if (linkNodes) {
          linkNodes.forEach((node) => {
            if (manager.DOMApis.isCssLinkNode(node)) {
              const href = manager.findAttributeValue(node, 'href');
              href &&
                safeLoad(
                  loader,
                  info.name,
                  baseUrl ? transformUrl(baseUrl, href) : href,
                  false,
                  immediately,
                );
            }
          });
        }
        if (metaNodes) {
          metaNodes.forEach((node) => {
            if (manager.DOMApis.isRemoteModule(node)) {
              const src = manager.findAttributeValue(node, 'src');
              if (src && isAbsolute(src)) {
                safeLoad(loader, info.name, src, true, immediately);
              } else {
                warn(
                  `The loading of the remote module must be an absolute path. "${src}"`,
                );
              }
            }
          });
        }
      }
    };
    if (immediately) {
      loadStaticResource();
    } else {
      requestIdleCallback(loadStaticResource);
    }
  });
}

export function getRanking() {
  const str = localStorage.getItem(storageKey);
  if (str) {
    const data = JSON.parse(str);
    return data.sort((a, b) => b.count - a.count);
  }
  return [];
}

export function setRanking(appName: string) {
  const str = localStorage.getItem(storageKey);
  const newCurrent = { appName, count: 1 };

  if (!str) {
    safeWrapper(() =>
      localStorage.setItem(storageKey, JSON.stringify([newCurrent])),
    );
  } else {
    const data = JSON.parse(str);
    const current = data.find((app) => app.appName === appName);
    current ? current.count++ : data.push(newCurrent);
    safeWrapper(() => localStorage.setItem(storageKey, JSON.stringify(data)));
  }
}

const loadedMap = Object.create(null); // Global cache, only load again is enough

declare module '@garfish/core' {
  export default interface Garfish {
    preload: (appName: string) => void;
  }
}

export function GarfishPreloadPlugin() {
  return function (Garfish: interfaces.Garfish): interfaces.Plugin {
    Garfish.preload = (appName) => {
      loadAppResource(Garfish.loader, Garfish.appInfos[appName], true);
    };

    return {
      name: 'preload',
      version: __VERSION__,

      beforeLoad(appInfo) {
        if (Garfish.options.disablePreloadApp) {
          return;
        }
        setRanking(appInfo.name);
      },

      registerApp(appInfos) {
        // Through disablePreloadApp preload is prohibited
        if (Garfish.options.disablePreloadApp) {
          return;
        }
        setTimeout(
          () => {
            if (isMobile || isSlowNetwork()) return;
            const ranking = getRanking();

            for (const { appName } of ranking) {
              if (appInfos[appName] && !loadedMap[appName]) {
                loadedMap[appName] = true;
                loadAppResource(Garfish.loader, appInfos[appName]);
              }
            }

            for (const key in appInfos) {
              if (!loadedMap[key]) {
                loadAppResource(Garfish.loader, appInfos[key]);
              }
            }
          },
          __TEST__ ? 0 : 5000,
        );
      },
    };
  };
}
