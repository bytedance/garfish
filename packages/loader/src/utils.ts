import {
  error,
  isPrimitive,
  isPlainObject,
  parseContentType,
} from '@garfish/utils';
import { Manager, Loader } from './index';

export async function request(url: string, config: RequestInit) {
  const result = await fetch(url, config || {});
  // Response codes greater than "400" are regarded as errors
  if (result.status >= 400) {
    error(`"${url}" load failed with status "${result.status}"`);
  }
  const code = await result.text();
  const type = result.headers.get('content-type');
  const mimeType = parseContentType(type);
  return { code, result, mimeType };
}

export function copyResult(result) {
  if (result.resourceManager) {
    result.resourceManager = (result.resourceManager as Manager).clone();
  }
  return result;
}

// Compatible with old api
export function mergeConfig(loader: Loader, url: string) {
  const extra = loader.requestConfig;
  const config = typeof extra === 'function' ? extra(url) : extra;
  return { mode: 'cors', ...config } as RequestInit;
}
