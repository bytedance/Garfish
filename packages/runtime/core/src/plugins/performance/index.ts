import SubAppObserver from './subAppObserver';
import { interfaces } from '../../index';

// 在 Garfish 对应的生命周期中进行关键节点注册
export function GarfishPerformance() {
  const subAppObserver = null;
  return function (_Garfish: interfaces.Garfish): interfaces.Plugin {
    const subAppMap = {};
    return {
      name: 'performance',
      async beforeLoad(appInfo: any) {
        if (!subAppMap[appInfo.name]) {
          // 运营平台子应用跟节点 id 为 'master-app'
          subAppMap[appInfo.name] = new SubAppObserver({
            subAppRootSelector: appInfo.domGetter as Element,
          });
        }

        // 运营平台使用 Slardar 自定义事件上报子应用性能数据
        // subAppObserver.subscribePerformanceData((subAppTimeData: IPerformanceData) => {
        //   // subAppTimeData: 订阅的当前子应用的性能数据，详情看 interface 定义
        //   // 子应用性能数据上报（根据业务或数据为度需求不同可自行调整）
        //   // 以下代码是运营平台对子应用性能数据的上报代码，仅作为示例！
        //   const report = name => {
        //     // window.Slardar('emit', 'timer', {
        //     //   name,
        //     //   value: subAppTimeData[name],
        //     //   tags: {
        //     //     path: subAppTimeData.path,
        //     //     subAppEntry: subAppTimeData.path,
        //     //     env: process.env.NODE_ENV,
        //     //     isTimeout: String(subAppTimeData[name] >= 5000)
        //     //   }
        //     // })
        //     console.log(name, subAppTimeData[name])
        //   };

        //   // 子应用资源加载时间 subAppTimeData.resourceLoadTime
        //   report('resourceLoadTime');
        //   // 子应用资源白屏时间 subAppTimeData.blankScreenTime
        //   report('blankScreenTime');
        //   // 子应用资源首屏时间 subAppTimeData.firstScreenTime
        //   report('firstScreenTime');
        // });
        subAppMap[appInfo.name].subAppBeforeLoad(appInfo.entry);
        return true;
      },
      afterLoad(appInfo, appInstance: any) {
        if (appInstance)
          appInstance.appPerformance = subAppMap[appInfo.name] as any;
      },
      beforeMount(appInfo: any) {
        subAppMap[appInfo.name].subAppBeforeMount(appInfo.entry);
      },
      beforeUnMount(appInfo: any) {
        subAppMap[appInfo.name].subAppUnMount(appInfo.entry);
      },
    };
  };
}
