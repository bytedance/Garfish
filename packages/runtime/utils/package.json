{
  "name": "@garfish/utils",
  "version": "0.0.32",
  "description": "utils module.",
  "keywords": [
    "garfish",
    "utils"
  ],
  "author": "zhouxiao <codingzx@gmail.com>",
  "homepage": "https://github.com/bytedance/garfish",
  "license": "Apache-2.0",
  "main": "index.js",
  "module": "dist/utils.esm-bundler.js",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/bytedance/garfish.git"
  },
  "bugs": {
    "url": "https://github.com/bytedance/garfish/issues"
  },
  "scripts": {},
  "types": "dist/index.d.ts",
  "buildOptions": {
    "name": "Utils",
    "devTemplate": "module",
    "formats": [
      "esm-bundler",
      "esm-browser",
      "cjs",
      "umd"
    ]
  },
  "peerDependencies": {
    "@garfish/core": "^0.0.23"
  },
  "publishConfig": {
    "registry": "https://registry.npmjs.org",
    "access": "public"
  },
  "gitHead": "4a882c7f2734d085689d423f5e0c51d9a0eccbba"
}
