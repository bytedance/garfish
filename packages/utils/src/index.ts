export * from './utils';
export * from './queue';
export * from './sentry';
export * from './domApis';
export * from './garfish';
export * from './mimeType';
export * from './container';
export * from './testResource';
export * from './templateParse';
