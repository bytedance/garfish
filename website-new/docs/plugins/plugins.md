---
title: Garfish 插件
slug: /garfish-plugins
order: 2
---

## Garfish 插件
