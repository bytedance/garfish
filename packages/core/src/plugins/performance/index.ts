import { getRenderNode } from '@garfish/utils';
import { interfaces } from '../../index';
import { SubAppObserver } from './subAppObserver';

// Key nodes in Garfish corresponding to the life cycle of registration
export function GarfishPerformance() {
  return function (): interfaces.Plugin {
    const subAppMap = {};
    return {
      name: 'performance',

      async beforeLoad(appInfo) {
        if (!subAppMap[appInfo.name]) {
          const appDomGetter = await getRenderNode(appInfo.domGetter);
          subAppMap[appInfo.name] = new SubAppObserver({
            subAppRootSelector: appDomGetter as Element,
          });
        }
        subAppMap[appInfo.name].subAppBeforeLoad(appInfo.entry);
        return true;
      },

      afterLoad(appInfo, appInstance: interfaces.App) {
        if (appInstance) {
          appInstance.appPerformance = subAppMap[appInfo.name] as any;
        }
      },

      beforeMount(appInfo) {
        subAppMap[appInfo.name].subAppBeforeMount(appInfo.entry);
      },

      beforeUnmount(appInfo) {
        subAppMap[appInfo.name].subAppUnmount(appInfo.entry);
      },
    };
  };
}
