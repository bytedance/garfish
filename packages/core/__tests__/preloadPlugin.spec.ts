import { mockStaticServer } from '@garfish/utils';
import Garfish from '../src/index';
import { GarfishPreloadPlugin, storageKey } from '../src/plugins/preload';

describe('Core: preload plugin', () => {
  const vueSubAppEntry = './resources/vueApp.html';
  const reactSubAppEntry = './resources/reactApp.html';

  mockStaticServer(__dirname);

  it('disablePreloadApp is true close setRanking', async () => {
    const lifecycle = GarfishPreloadPlugin()({
      options: { disablePreloadApp: true },
    } as Garfish);
    lifecycle.beforeLoad({ name: 'vue-app' });

    expect(localStorage.getItem(storageKey)).toBeNull();
  });

  it('disablePreloadApp is false use setRanking', async () => {
    const lifecycle = GarfishPreloadPlugin()({
      options: { disablePreloadApp: false },
    } as Garfish);
    lifecycle.beforeLoad({ name: 'vue-app' });
    lifecycle.beforeLoad({ name: 'vue-app' });
    lifecycle.beforeLoad({ name: 'react-app' });

    expect(localStorage.getItem(storageKey)).toMatchSnapshot();
  });
});
