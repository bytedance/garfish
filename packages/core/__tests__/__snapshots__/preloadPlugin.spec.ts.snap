// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Core: preload plugin disablePreloadApp is false use setRanking 1`] = `null`;
